jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala

```diff
@@ -0,0 +1,371 @@
+        .take(expectedSize)
+        .map(env => (env, env.message.asInstanceOf[TextMessage].getText))
+        .map { case (env, text) => env.acknowledge(); text }
+        .runWith(Sink.seq)
+        .map(_.sorted)
+      //#run-topic-source
+
+      //We wait a little to be sure that the source is connected
+      Thread.sleep(500)
+
+      //#run-topic-sink
+      Source(in).runWith(jmsTopicSink)
+      //#run-topic-sink
+      Source(inNumbers).runWith(jmsTopicSink2)
+
+      val expectedList: List[String] = in ++ inNumbers
+      result1.futureValue should contain theSameElementsAs expectedList
+      result2.futureValue should contain theSameElementsAs expectedList
+    }
+
+    "ensure no message loss when stopping a stream" in withServer() { ctx =>
+      val url: String = ctx.url
+      val connectionFactory = new ActiveMQConnectionFactory(url)
+
+      val jmsSink: Sink[JmsTextMessage, NotUsed] = JmsSink(
+        JmsSinkSettings(connectionFactory).withQueue("numbers")
+      )
+
+      val numsIn = 1 to 100
+      val msgsIn = numsIn map { n =>
+        JmsTextMessage(n.toString).add("Number", n)
+      }
+
+      Source(msgsIn).runWith(jmsSink)
+
+      val jmsSource: Source[AckEnvelope, KillSwitch] = JmsSource.ackSource(
+        JmsSourceSettings(connectionFactory).withSessionCount(5).withBufferSize(0).withQueue("numbers")
+      )
+
+      val resultQueue = new LinkedBlockingQueue[String]()
+
+      val killSwitch = jmsSource
+        .take(msgsIn.size)
+        .throttle(10, 1.second, 2, ThrottleMode.shaping)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
+
+      Thread.sleep(2000)
+      killSwitch.shutdown()
+
+      Thread.sleep(500)
+      println("Elements in resultQueue before first shutdown: " + resultQueue.size)
+
+      val killSwitch2 = jmsSource
+        .take(msgsIn.size - resultQueue.size)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
+
+      val resultList = new mutable.ArrayBuffer[String](numsIn.size)
+
+      @tailrec
+      def keepPolling(): Unit =
+        Option(resultQueue.poll(2, TimeUnit.SECONDS)) match {
+          case Some(entry) =>
+            resultList += entry
+            keepPolling()
+          case None =>
+        }
+
+      keepPolling()
+
+      println("Elements in resultList now: " + resultList.size)
+      killSwitch2.shutdown()
+
+      resultList should contain theSameElementsAs numsIn.map(_.toString)
+    }
+
+    "ensure no message loss when aborting a stream" in withServer() { ctx =>
+      val url: String = ctx.url
+      val connectionFactory = new ActiveMQConnectionFactory(url)
+
+      val jmsSink: Sink[JmsTextMessage, NotUsed] = JmsSink(
+        JmsSinkSettings(connectionFactory).withQueue("numbers")
```

Comment:
Isn't there a risk that the tests interfere with each other if we use the same queue in many test cases?

Reply:
For all the tests we have, each test spawns its separate broker. There is nothing persisted so each test is really separate. For what I do for our own use cases of this, I start one broker per test suite. In that case using the same queue may have side-effects (depending on how clean you leave it.

The barebones ActiveMQ broker is light enough to keep going with this scheme. Fixing this would need restructuring the tests - even the ones already there. I'm happy to help, but I feel this needs a separate issue.